### PR TITLE
feat:  Design and implement several variable charts

### LINKED ISSUE #654: [OSPP 2023] Project 2: Implement the new "Repo Collections" feature (only charts)
## Difficulty
Advance

## Mentor
@zhicheng-ning

### PR DESCRIPTION
**Pull request by wj23027**
## Brief Information

This pull request is in the type of ([more info](https://github.com/angular/angular/blob/main/CONTRIBUTING.md#type) about types):

- [ ] build
- [ ] ci
- [ ] docs
- [x] feat
- [ ] fix
- [ ] perf
- [ ] refactor
- [ ] test

Related issues ([all available keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

- resolve #654 

## Details

I have implemented a basic grid layout for the charts, allowing different-sized charts to be added to the layout. 

image

In the next steps, I will need to add logic to allow users to add charts of specific sizes and types to specific positions, and optimize the styling of the layout.

Next Steps:
- [ ] Adjust the visual presentation of the charts.
- [ ] Implement Additional Chart Styles
- [ ] optimize the styling of the layout
- [ ] add logic to allow users to add charts of specific sizes and types to specific positions
- [ ] i18n
- [ ] Integrate the feature into 'Repo Collections' feature
## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/hypertrons/hypertrons-crx/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/hypertrons/hypertrons-crx)

### CODE REVIEW

**Comment by andyhuang18** [comment]
Great! Wish you have a good time in your trip. 😄

**Comment by wxharry** [comment]
Hi, @wj23027 . I have changed this base branch to `feat/repo-collection` since you and Andy is working on one feature. I have implmented a modal view for testing new charts. It helps to focus on developing the component itself. Since this branch highly depends on the work from Andy, my suggestion is that you can work on new charts first, wait til Andy implements the modal.

**Review by tyn1998 — CHANGES_REQUESTED** [review]
Hi @wj23027, I found that the GridLayout is not imported by any file so I guess the screenshot you provided in the PR description is just for preview. Since @wxharry [provided a way ](https://github.com/hypertrons/hypertrons-crx/pull/716#issuecomment-1690978483)to test your charts, you can import GridLayout there.

Besides, please comment the code in English and hurry up :D

**Comment by wj23027** [comment]
This update introduces two new charts: boxplot chart and a Sankey chart . The box plot is used to illustrate the approximate distribution of issue response times, while the Sankey chart showcases user activity across different repositories.

boxplot chart :

image

Sankey chart：

image

**Comment by wj23027** [comment]
In addition, a simple grid layout component has been implemented to display the charts separately in both the modal and option pages.

image

Furthermore, clicking on menu items can highlight data for corresponding repositories in certain charts.



Next Tasks :
- Implement highlighting for other charts.
- Fine-tune chart styles.
- Merge Andy's code contributions.

**Comment by tyn1998** [comment]
@wj23027 The Sankey chart is really good.

There are only two repos in the demo collection, could you add more repos into it? Please bare in mind that these charts should be scalable for any given collection of different size.

**Comment by tyn1998** [comment]
> @wj23027 The Sankey chart is really good.
> 
> There are only two repos in the demo collection, could you add more repos into it? Please bare in mind that these charts should be scalable for any given collection of different size.

Hello @wj23027, could you add more repos into your mock data? Since in my view, these two charts are not a good choice for those collections with 3+ repos.

image

File: src/pages/Options/variable-charts/GridLayout.tsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        grid.push(
+          <div
+            key={index}
+            className={`grid-item size-${chartSize[size]}`}
+            onClick={() => setGridItemSize(index, newsize)}
+          >
+            {/* 图表组件，根据占用情况来决定显示的内容 */}
+            {size > 0 ? <LineChart /> : ''}
+          </div>
+        );
```
Comment (tyn1998) [inline src/pages/Options/variable-charts/GridLayout.tsx:42]: Hello @wj23027, There is an error in line#40, if you `yarn run start` you'll find it.

image

File: src/helpers/get-newest-month.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+  const lastDataAvailableMonth = new Date();
+  lastDataAvailableMonth.setDate(0);
+  return (
+    lastDataAvailableMonth.getFullYear() +
+    '-' +
+    (lastDataAvailableMonth.getMonth() + 1).toString().padStart(2, '0')
+  );
+};
+
+export default getNewestMonth;
```
Comment (tyn1998) [inline src/helpers/get-newest-month.ts:11]: This way to get the lastDataAvailableMonth is not robust. A solid way is implemented in #635, which you can refer to :)

**Review by tyn1998 — CHANGES_REQUESTED** [review]
Hi @wj23027, thanks for your time and effort contributed to this OSPP project. Here are some suggestions for you and I will give more of them in the following days.

**Comment by wj23027** [comment]
In this update, I added numerical display panels.I tried increasing the number of repositories in the collection and noticed that after the increase, the code additions and deletions chart as well as the multi-dimensional bar chart didn't look very aesthetically pleasing. So, I made the following adjustments: 
1.  Changed the code additions and deletions chart to an area chart,
2.  Limited the display range of the multi-dimensional bar chart to the last 6 months, 
3.  Added stacked bar chart.

**Comment by wj23027** [comment]
Here is a screenshot of the updated charts:

image

image

**Comment by wj23027** [comment]
Additionally, I've noticed that when the collection contains too many developers, there are some issues with the Sankey chart's display. Therefore, I've increased the size of the Sankey chart, but it still works best for collections containing no more than 150 developers (ideally, fewer than 100).

File: src/helpers/get-newest-month.ts (hunk 2 above)
Comment (wj23027) [reply src/helpers/get-newest-month.ts]: Thank you for pointing it out! I've made the necessary changes!

File: src/pages/Options/variable-charts/GridLayout.tsx (hunk 1 above)
Comment (wj23027) [reply src/pages/Options/variable-charts/GridLayout.tsx]: The error likely resulted from missing required parameters. It's a deprecated file, which I've now removed.

**Comment by wj23027** [comment]
I have an issue to address: in the [stacked bar chart](https://github.com/hypertrons/hypertrons-crx/pull/716/files#diff-58ce18998c0f216466c14d33a617a8ff12ea43ea29d3d13e5978cc33283d8776), the data from different series isn't stacking as intended, but I haven't found the reason yet.

image

**Comment by andyhuang18** [comment]
> Here is a screenshot of the updated charts:
> 
> image
> 
> image

Looks good!

**Comment by wxharry** [comment]
> Additionally, I've noticed that when the collection contains too many developers, there are some issues with the Sankey chart's display. Therefore, I've increased the size of the Sankey chart, but it still works best for collections containing no more than 150 developers (ideally, fewer than 100).

Great! I am glad you tested it. Then we should add a warning or notice like when users are trying to add a sankey chart. I prefer adding it anyway since this is a very cool chart. What do you think @tyn1998 ?



> I have an issue to address: in the [stacked bar chart](https://github.com/hypertrons/hypertrons-crx/pull/716/files#diff-58ce18998c0f216466c14d33a617a8ff12ea43ea29d3d13e5978cc33283d8776), the data from different series isn't stacking as intended, but I haven't found the reason yet.
> 

I think it is because the two arries of data have different length.
image

**Comment by wxharry** [comment]
Also, please pay attention to the style below. The charts are not supposed to be outside of their containers. I hope it's a small fix, but if it's not, a temporary workaround is acceptable.
image

**Comment by zhicheng-ning** [comment]
Some suggestions：
- The dimensions of the chart need to be adapted. For example：change it from150000 to 150k
- The maximum value of the chart is adapted,it will be more beautiful. For example：change maximun value from 120 to 20（The maximum value among all existing values）.

> Also, please pay attention to the style below. The charts are not supposed to be outside of their containers. I hope it's a small fix, but if it's not, a temporary workaround is acceptable. image

**Comment by wj23027** [comment]
Hi, I merged the collection contents(including the repolist and dashboard) into the Collection Modal~

image

**Review by tyn1998 — APPROVED** [review]
Great job @wj23027  !

Your work LGTM, thank you for your contributions!

**Pull request merged**